pkg/agent/flowexporter/priorityqueue/priorityqueue.go

```diff
@@ -0,0 +1,143 @@
+// distributed under the License is distributed on an “AS IS” BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package priorityqueue
+
+import (
+	"container/heap"
+	"time"
+
+	"antrea.io/antrea/pkg/agent/flowexporter"
+)
+
+var minExpiryTime = 100 * time.Millisecond
+
+type ExpirePriorityQueue struct {
+	items             []*flowexporter.ItemToExpire
+	ActiveFlowTimeout time.Duration
+	IdleFlowTimeout   time.Duration
+	KeyToItem         map[flowexporter.ConnectionKey]*flowexporter.ItemToExpire
+}
+
+func NewExpirePriorityQueue(activeFlowTimeout time.Duration, idleFlowTimeout time.Duration) *ExpirePriorityQueue {
+	return &ExpirePriorityQueue{
+		items:             make([]*flowexporter.ItemToExpire, 0),
+		ActiveFlowTimeout: activeFlowTimeout,
+		IdleFlowTimeout:   idleFlowTimeout,
+		KeyToItem:         make(map[flowexporter.ConnectionKey]*flowexporter.ItemToExpire),
+	}
+}
+
+func (pq *ExpirePriorityQueue) Len() int {
+	return len(pq.items)
+}
+
+func (pq *ExpirePriorityQueue) minExpireTime(i int) time.Time {
+	if pq.items[i].ActiveExpireTime.Before(pq.items[i].IdleExpireTime) {
+		return pq.items[i].ActiveExpireTime
+	} else {
+		return pq.items[i].IdleExpireTime
+	}
+}
+
+func (pq *ExpirePriorityQueue) Less(i, j int) bool {
+	return pq.minExpireTime(i).Before(pq.minExpireTime(j))
+}
+
+func (pq *ExpirePriorityQueue) Swap(i, j int) {
+	pq.items[i], pq.items[j] = pq.items[j], pq.items[i]
+	pq.items[i].Index = i
+	pq.items[j].Index = j
+}
+
+func (pq *ExpirePriorityQueue) Push(x interface{}) {
+	n := len((*pq).items)
+	item := x.(*flowexporter.ItemToExpire)
+	item.Index = n
+	(*pq).items = append((*pq).items, item)
+}
+
+func (pq *ExpirePriorityQueue) Pop() interface{} {
+	n := len((*pq).items)
+	item := ((*pq).items)[n-1]
+	item.Index = -1
+	(*pq).items = ((*pq).items)[0:(n - 1)]
+	return item
+}
+
+// Peek returns the item at the beginning of the queue, without removing the
+// item or otherwise mutating the queue. It is safe to call directly.
+func (pq *ExpirePriorityQueue) Peek() *flowexporter.ItemToExpire {
+	if pq.Len() == 0 {
+		return nil
+	}
+	return pq.items[0]
+}
+
+// Update modifies the priority of an Item in the queue.
+func (pq *ExpirePriorityQueue) Update(item *flowexporter.ItemToExpire, activeExpireTime time.Time, idleExpireTime time.Time) {
+	item.ActiveExpireTime = activeExpireTime
+	item.IdleExpireTime = idleExpireTime
+
+	heap.Fix(pq, item.Index)
+}
+
+// GetExpiryFromExpirePriorityQueue returns the shortest expire time duration
+// from expire priority queue.
+func (pq *ExpirePriorityQueue) GetExpiryFromExpirePriorityQueue() time.Duration {
+	currTime := time.Now()
+	if pq.Len() > 0 {
+		// Get the minExpireTime of the top item in ExpirePriorityQueue.
+		expiryDuration := minExpiryTime + pq.minExpireTime(0).Sub(currTime)
+		if expiryDuration < 0 {
+			return minExpiryTime
+		}
```

Comment:
this is weird because it seems that in some cases we can return a value which is less than `minExpiryTime`

I would expect the code to look like this:
```go
expiryDuration := pq.minExpireTime(0).Sub(currTime)
if expiryDuration < minExpiryTime {
    return minExpiryTime
}
```

Reply:
> this is weird because it seems that in some cases we can return a value which is less than `minExpiryTime`

Yes, we can return a value which is less than `minExpiryTime`.
The const is used to create "batch-processing". We want to process multiple records at one time, if their expiry time are very close(within 100ms). 
I copied this method from flow-aggregator's priority queue implementation, related discussion: https://github.com/vmware/go-ipfix/pull/185#discussion_r629790308,  https://github.com/vmware/go-ipfix/pull/185#discussion_r630610796

> I would expect the code to look like this:

As Srikar mentioned in the link, this will not cover the case where the top item's expiry is > 100ms, and the second item's expiry is very close to 100ms, e.g. 110ms.

The variable name is a bit mis-leading. I added a comment for it.

Reply:
ok, I remember discussing this with Srikar in the past